### ISSUE TITLE
Intermediate inputs for sequential blocks

### ISSUE THREAD

**Issue by BenjaminMidtvedt** [DeepTrackAI/deeplay#14]
Continuation from #13 

My suggestion would to have a subclassed

class GraphSequentialBlock(SequentialBlock):
     ....

where the interleaving of inputs is handled in the forward call. @JesusPinedaC thoughts?

**Comment by BenjaminMidtvedt** [comment]
I guess there could be some ambiguity in the forward call for what inputs the layer wants

**Comment by JesusPinedaC** [comment]
I have come up with the following proposal:

As you suggested, we build upon `SequentialBlock`:

```python
class BranchedSequentialBlock(SequentialBlock):
    """
    Extension of SequentialBlock that supports extra inputs.
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def forward(self, x, *args: ExtraInput):
        """
        Perform forward pass through the branched block based on ExtraInput instances.

        Args:
            x: The input data.
            *args (ExtraInput): Variable number of ExtraInput instances.

        Returns:
            The output of the forward pass.

        Example:
            # Forwarding data through the branched block
            input_data = ...  # Your input data here
            result = branched_block.forward(input_data, extra_input1, extra_input2, ...)

        """
        for name in self.order:
            extra_inputs = self.assess_extra_inputs(args, name)
            x = getattr(self, name)(x, *extra_inputs["args"], **extra_inputs["kwargs"])
        return x

    def assess_extra_inputs(self, args, name):
        """
        Assess and organize ExtraInput instances based on the target name.

        Args:
            args (Iterable[ExtraInput]): Iterable of ExtraInput instances.
            name (str): The name of the current layer.

        Returns:
            Dictionary containing lists of arguments and keyword arguments.
        """
        matching_targets = filter(lambda arg: arg.target == name, args)

        extra_inputs = {"args": [], "kwargs": {}}
        for target in matching_targets:
            if target.name is None:
                extra_inputs["args"].append(target.value)
            else:
                extra_inputs["kwargs"][target.name] = target.value

        return extra_inputs
```

Similarly to `SequentialBlock`, `BranchedSequentialBlock` receives the core data yet allows for extra inputs. Extra inputs are explicitly defined as ExtraInput instances, a particular data type defined by its value, target layer, and, optionally, a name. Internally, BranchedSequentialBlock matches each extra input with its corresponding target layer.

Consider a block consisting of a linear layer, followed by Relu activation and Layer normalization. 

```python
block = BranchedSequentialBlock(
    layer=torch.nn.LazyLinear(10),
    activation=torch.nn.ReLU(),
    normalization=tg.nn.norm.LayerNorm(10),
)
```
The core data consists of a `100x2` array where elements in the batch are distributed along dim 0. This does not affect operations in `layer` and `activation`. However, `normalization` (in this case, Layer normalization) will require extra information on the batch indices, i.e., 

```python
x = torch.randn(100, 10)
batch_index = torch.randint(0, 2, (100,))

batch_index = ExtraInput(value=batch_index, target="normalization")
out = block(x, batch_index)
```

**Comment by JesusPinedaC** [comment]
@BenjaminMidtvedt thoughts?

**Comment by BenjaminMidtvedt** [comment]
@JesusPinedaC I think you're on the right track. However, I think the ExtraInput object should be a part of the Block, or maybe even a part of the Layer. The reason is that otherwise it's not really configurable. If this block is deep in a network, I can't on the surface change what ExtraInputs are passed to the block.

**Comment by BenjaminMidtvedt** [comment]
Block could be called with something like "ExtraInput", which is available to every Layer. Then what you configure is which layer takes what from the extra. So, 
```py
block = BranchedSequentialBlock(
    layer=torch.nn.LazyLinear(10),
    activation=torch.nn.ReLU(),
    normalization=tg.nn.norm.LayerNorm(10),
)
blocks.normalization.set_inputs("x", "batch_index")

# can make it a fancier object
extras = dict(batch_index=...)

block(x, extras)
```

**Comment by JesusPinedaC** [comment]
@BenjaminMidtvedt Great!

**Comment by JesusPinedaC** [comment]
So something like this could work:

```python
from .sequential import SequentialBlock

from typing import Optional
from functools import partial



class BranchedSequentialBlock(SequentialBlock):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def forward(self, x, extras: Optional[dict] = {}):
        _extras = {}
        for name in self.order:
            layer = getattr(self, name)

            # Check if there are any extra inputs for this layer
            layer_extras = [(le, extras.get(le)) for le in getattr(layer, "extras", [])]
            _extras.update(layer_extras)

            x = getattr(self, name)(x, **_extras)
        return x

    def __getattr__(self, name):
        if name in self.order:
            layer = super().__getattr__(name)
            if not hasattr(layer, "set_extras"):
                layer.set_extras = partial(
                    lambda self, *args: setattr(self, "extras", list(args)), layer
                )
            return layer
        raise AttributeError(
            f"'{type(self).__name__}' object has no attribute '{name}'"
        )
```

Example:

```python
block = BranchedSequentialBlock(
    layer=torch.nn.LazyLinear(10),
    activation=torch.nn.ReLU(),
    normalization=tg.nn.norm.LayerNorm(10),
)
block.normalization.set_extras("batch")

x = torch.randn(100, 10)
extras = dict(batch=torch.randint(0, 2, (100,)))

output = block(x, extras)
```

What do you think @BenjaminMidtvedt ?

**Comment by BenjaminMidtvedt** [comment]
I think my only remaining concern is that we need to also allow:

set_extras(argname="otherargname") for cases where the name in the layer is different for what is in the block. The rest is implementation details we can sort out in a PR

**Comment by JesusPinedaC** [comment]
Solved by https://github.com/DeepTrackAI/deeplay/pull/28
